### PR TITLE
`pupa` templating & rollup powered build

### PR DESCRIPTION
**Pull request by vladimyr**
- generating `stats.html` from mustache template
- building plugin with rollup (leveraging `rollup-plugin-postcss`); removing gulp
- moving away from experimental `fs.promises` api; using `util.promisify` instead
- updating `package.json` (adding `homepage` & `bugs`) fields
- updating eslint configuration
- refactoring uid module
- moving sass to `src/style` dir

### CODE REVIEW

File: .eslintrc.yml [hunk 1]
```diff
@@ -9,29 +8,28 @@ parserOptions:
...
-  strict: [2, "global"]
-  semi: [2, "always"]
-  no-sequences: 2
-  dot-notation: 1
-  dot-location: [2, "property"]
-  global-require: 2
-  no-path-concat: 2
-  prefer-arrow-callback: 2
-  no-var: 2
+  quotes: ["error", "double"]
```
Comment (vladimyr) [inline .eslintrc.yml]: I've turned on this because you almost exclusively used double quotes through whole codebase.

File: src/.eslintrc.yml [hunk 2]
```diff
@@ -1,8 +1,4 @@
-extends:
-  - ../.eslintrc.yml
```
Comment (vladimyr) [inline src/.eslintrc.yml]: This is not needed, eslint will automatically extend upper one.

**Comment by vladimyr** [comment]
@btd Please ignore whitespace changes while reviewing this by appending `?w=1` to PR url: https://github.blog/2018-05-01-ignore-white-space-in-code-review/

Those are CRLF to LF modifications which my editor automatically does according to `.editorconfig` rules. I don't know which editor you use but it shouldn't take more than 5 secs to properly set it up: https://editorconfig.org/#download

File: rollup.config-dev.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"use strict";
+
+const baseConfig = require("./rollup.config");
+const plugin = require("./plugin");
+
+module.exports = baseConfig.map((options, index) => {
+  options.plugins.push(plugin({
+    open: true,
+    filename: `stats.${index}.html`,
+    template: options.$template
```
Comment (btd) [inline rollup.config-dev.js]: Are you sure there should be `$template`? and not just `template`?

File: package.json [hunk 4]
```diff
@@ -32,21 +40,16 @@
...
     "rollup-plugin-commonjs": "^10.0.0",
-    "rollup-plugin-node-resolve": "^5.0.1",
+    "rollup-plugin-node-resolve": "^5.0.3",
+    "rollup-plugin-postcss": "^2.0.3",
     "rollup-plugin-uglify": "^6.0.2",
-    "typeface-oswald": "0.0.54",
-    "vinyl-buffer": "^1.0.1",
-    "vinyl-source-stream": "^2.0.0",
-    "yargs": "^13.2.4"
+    "sass": "^1.21.0",
```
Comment (btd) [inline package.json]: :+1: i actually don't know they have js version.

File: src/uid.js [hunk 5]
```diff
@@ -1,14 +1,27 @@
...
+  get href() {
+    const url = new URL(window.location);
+    url.hash = this.id;
+    return url.href;
+  }
+
+  toString() {
+    return `url(${this.href})`;
+  }
+}
```
Comment (btd) [inline src/uid.js]: Previous version of id has cached version of href.

File: build-stats.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+"use strict";
+
+const Mustache = require("mustache");
+const path = require("path");
+const readFile = require("util").promisify(require("fs").readFile);
```
Comment (btd) [inline build-stats.js]: Please move require(fs) to the top.

**Comment by btd** [comment]
Thank you for PR, it is very good. Actually i learned several things about rollup. There are several minor comments and actually 2 things i would like to be changed:
1. I hate mustache (this is imo). I also thought about using basic template, but i thought to use lodash.template, or just call escaping where necessary. It would be very nice if you would change to use it (and it think it is ok to rest template as js string, there is noting bad about it).
2. Refactoring of uid, is very subjective (your version of .href worse). I would rest class, but set .id and .href as properties as they were.

File: package.json (hunk 4 above)
Comment (vladimyr) [reply package.json]: Actually it is written in dart but transpiled back to js. It is now official sass implementation. 😎

File: src/uid.js (hunk 5 above)
Comment (vladimyr) [reply src/uid.js]: True that, failed to figure out that is actually important :upside_down_face: I'll cache it again.

File: build-stats.js (hunk 6 above)
Comment (vladimyr) [reply build-stats.js]: Ok :bowtie:

**Comment by vladimyr** [comment]
> Thank you for PR, it is very good. Actually i learned several things about rollup.

Yay :tada: 

> 1. I hate mustache (this is imo). I also thought about using basic template, but i thought to use lodash.template, or just call escaping where necessary. It would be very nice if you would change to use it (and it think it is ok to rest template as js string, there is noting bad about it).

I don't know why you hate `mustache` but `_.template` isn't good alternative either. Mustache is logicless which inherently prevents you from making messy templates and kinda increases security because it doesn't support arbitrary js evaluation like `_.template` or template strings. (See more here: https://github.com/jashkenas/underscore/issues/1898)
Maybe this could be our sweet spot between `{` and template strings: https://github.com/sindresorhus/pupa ? It has escaping and does not allow js code execution inside template?

> 2\. Refactoring of uid, is very subjective (your version of .href worse). I would rest class, but set .id and .href as properties as they were.

I'll rework that :construction:

**Comment by btd** [comment]
I like idea of `pupa`, lets try it. 

Thank you again for great PR.

**Comment by btd** [comment]
Thank you!

**Pull request merged**